docs/proposals-accepted/202206-block-series-deletion-compactor.md

```diff
@@ -0,0 +1,206 @@
+
+There are two approaches about how to scan block indexes:
+1. Download all the blocks locally and check against the tombstones.
+2. Use part of the store gateway code to scan the block index and chunks on the object storage to generate `MemTombstones` result.
+
+Approaches 2 should be better as only block indexes need to be downloaded, chunks data are reading from the object storage on the fly so disk space can be saved. As compactor is an offline component, it is acceptable to read from the object storage rather than downloading the whole data.
+
+The pitfall is that we need to expose part of the store gateway code like `BucketIndexReader`, `BucketChunkReader` as well as other internal code to be used by the compactor.
+
+#### Option 2: New API to Match Tombstones
+
+Since the block indexes are already downloaded by the store gateway, it can already scan the blocks and calculate the tombstones. So a way here is to expose some APIs from store gateway, and then compactor can query the APIs to know what blocks can be deleted.
+
+The gRPC API is called `Tombstone` API. The compactor can either query from the store gateway directly or query from queriers, and queriers do the API federation from multiple store gateway instances.
+
+The example gRPC format looks like below. The response is almost the same as the `MemTombstone` result.
+
+```protobuf
+service Tombstone {
+  rpc Tombstone(TombstoneRequest) returns (TombstoneResponse);
+}
+
+message TombstoneRequest {
+  int64 min_time = 1;
+  int64 max_time = 2;
+  repeated LabelMatcher matchers = 3 [(gogoproto.nullable) = false];
+}
+
+message TombstoneResponse {
+  repeated SeriesIntervals series_intervals = 1;
+}
+
+message SeriesIntervals {
+  int64 series_ref = 1;
+  repeated Interval intervals = 2;
+}
+
+message Interval {
+  int64 start = 1;
+  int64 end = 2;
+}
+```
+
+After getting the query result, the compactor does almost the same thing as the previous option: generate the tombstone file during compaction time.
+
+The drawback for this approach is that we are inventing a new API for this purpose only, which might be a bit of pain in the future.
+
+#### Option 3: Delete Series using Rewrite Logic
+
+The previous two options do the series deletion at compaction time, this option does actual series deletion using block rewrite. For the index scanning, it is still required to choose either option 1 or 2 to avoid unncessary rewrite even if the block doesn't have any matched series.
+
+This is not so efficient as rewrite is a different process than compaction. We need to add this step and make sure rewrite can work together with compaction and no redundant work is done.
+
+Overall, among the 3 options, option 1 is better as it is easier to implement and straightforward.
```

Comment:
Same comment here about giving more evidence to your recommendation by putting it under a `My recommendation` (or something like this) heading. This is buried inside the description of an option and it's not part of the option itself.